File: Backend/src/main/typescript/safetrack-client.ts
export type UUID = string;

export interface User {
  id: UUID;
  name: string;
  email: string;
  avatar_url?: string;
}

export interface AuthResponse {
  token: string;
  user: User;
}

export interface Incident {
  id: UUID;
  user_id: UUID;
  type: 'accident' | 'hazard' | 'crime' | 'weather' | 'other';
  title: string;
  description?: string;
  latitude: number;
  longitude: number;
  severity: 'low' | 'medium' | 'high' | 'critical';
  status: 'active' | 'resolved' | 'investigating';
  media_url?: string;
  reporter_name?: string;
  created_at: string;
  updated_at: string;
}

export interface SavedRoute {
  id: UUID;
  user_id: UUID;
  name: string;
  origin_name?: string;
  destination_name?: string;
  origin_lat: number;
  origin_lng: number;
  destination_lat: number;
  destination_lng: number;
  route_data?: unknown;
  is_favorite: boolean;
  created_at: string;
}

export interface ChatResult {
  text: string;
  action: unknown;
}

export interface MusicTrack {
  id: UUID;
  user_id: UUID;
  title: string;
  artist?: string;
  album?: string;
  duration?: number;
  file_url: string;
  cover_url?: string;
  created_at: string;
}

export interface Playlist {
  id: UUID;
  user_id: UUID;
  name: string;
  cover_url?: string;
  track_count?: number;
  created_at: string;
}

export interface Ad {
  id: UUID;
  user_id?: UUID;
  business_name: string;
  description?: string;
  latitude: number;
  longitude: number;
  radius_km: number;
  image_url?: string;
  website_url?: string;
  payment_status?: 'pending' | 'paid' | 'expired';
  stripe_payment_intent_id?: string;
  active?: boolean;
  duration_days?: number;
  expires_at?: string;
  created_at?: string;
  distance_km?: number;
}

export class SafeTrackClient {
  constructor(private readonly baseUrl = 'http://localhost:4000', private token?: string) {}

  setToken(token: string) { this.token = token; }

  register(body: { name: string; email: string; password: string }) { return this.json<AuthResponse>('/api/auth/register', { method: 'POST', body }); }
  login(body: { email: string; password: string }) { return this.json<AuthResponse>('/api/auth/login', { method: 'POST', body }); }
  health() { return this.json<{ status: string; timestamp: string }>('/api/health'); }

  incidents() { return this.json<Incident[]>('/api/incidents'); }
  createIncident(form: FormData) { return this.form<Incident>('/api/incidents', form); }
  updateIncident(id: UUID, status: Incident['status']) { return this.json<Incident>(`/api/incidents/${id}`, { method: 'PATCH', body: { status } }); }
  deleteIncident(id: UUID) { return this.json<{ success: boolean }>(`/api/incidents/${id}`, { method: 'DELETE' }); }

  routes() { return this.json<SavedRoute[]>('/api/routes'); }
  saveRoute(body: Partial<SavedRoute>) { return this.json<SavedRoute>('/api/routes', { method: 'POST', body }); }
  toggleFavorite(id: UUID) { return this.json<SavedRoute>(`/api/routes/${id}/favorite`, { method: 'PATCH' }); }
  deleteRoute(id: UUID) { return this.json<{ success: boolean }>(`/api/routes/${id}`, { method: 'DELETE' }); }

  chat(message: string, history: Array<{ role: string; content: string }> = []) { return this.json<ChatResult>('/api/ai/chat', { method: 'POST', body: { message, history } }); }
  chatHistory() { return this.json<Array<{ role: string; content: string; created_at: string }>>('/api/ai/history'); }
  clearChatHistory() { return this.json<{ success: boolean }>('/api/ai/history', { method: 'DELETE' }); }

  tracks() { return this.json<MusicTrack[]>('/api/music/tracks'); }
  uploadTrack(form: FormData) { return this.form<MusicTrack>('/api/music/tracks', form); }
  deleteTrack(id: UUID) { return this.json<{ success: boolean }>(`/api/music/tracks/${id}`, { method: 'DELETE' }); }
  playlists() { return this.json<Playlist[]>('/api/music/playlists'); }
  createPlaylist(name: string) { return this.json<Playlist>('/api/music/playlists', { method: 'POST', body: { name } }); }
  addTrackToPlaylist(id: UUID, track_id: UUID, position = 0) { return this.json<{ success: boolean }>(`/api/music/playlists/${id}/tracks`, { method: 'POST', body: { track_id, position } }); }
  playlistTracks(id: UUID) { return this.json<MusicTrack[]>(`/api/music/playlists/${id}/tracks`); }

  ads() { return this.json<Ad[]>('/api/ads'); }
  nearbyAds(lat: number, lng: number) { return this.json<Ad[]>(`/api/ads/nearby?lat=${lat}&lng=${lng}`); }
  myAds() { return this.json<Ad[]>('/api/ads/mine'); }
  createAd(body: Partial<Ad>) { return this.json<Ad>('/api/ads', { method: 'POST', body }); }
  checkoutAd(id: UUID) { return this.json<Record<string, unknown>>(`/api/ads/${id}/checkout`, { method: 'POST' }); }
  activateAd(id: UUID) { return this.json<Ad>(`/api/ads/${id}/activate`, { method: 'POST' }); }
  deleteAd(id: UUID) { return this.json<{ success: boolean }>(`/api/ads/${id}`, { method: 'DELETE' }); }

  private async json<T>(path: string, init: { method?: string; body?: unknown } = {}): Promise<T> {
    const res = await fetch(`${this.baseUrl}${path}`, {
      method: init.method ?? 'GET',
      headers: { ...this.authHeader(), ...(init.body ? { 'Content-Type': 'application/json' } : {}) },
      body: init.body ? JSON.stringify(init.body) : undefined,
    });
    return this.parse<T>(res);
  }

  private async form<T>(path: string, body: FormData): Promise<T> {
    const res = await fetch(`${this.baseUrl}${path}`, { method: 'POST', headers: this.authHeader(), body });
    return this.parse<T>(res);
  }

  private authHeader(): HeadersInit { return this.token ? { Authorization: `Bearer ${this.token}` } : {}; }

  private async parse<T>(res: Response): Promise<T> {
    const data = await res.json();
    if (!res.ok) throw Object.assign(new Error(data.error ?? 'Request failed'), { status: res.status, data });
    return data as T;
  }
}

File: Backend/src/main/java/com/pathtocodequest/safetrack/SafeTrackClient.java
package com.pathtocodequest.safetrack;

import org.springframework.beans.factory.annotation.Value;
import org.springframework.http.HttpEntity;
import org.springframework.http.HttpHeaders;
import org.springframework.http.HttpMethod;
import org.springframework.http.MediaType;
import org.springframework.http.ResponseEntity;
import org.springframework.stereotype.Component;
import org.springframework.util.LinkedMultiValueMap;
import org.springframework.util.MultiValueMap;
import org.springframework.web.client.RestTemplate;
import org.springframework.web.client.HttpClientErrorException;
import com.fasterxml.jackson.annotation.JsonProperty;
import java.util.List;
import java.util.Map;

/**
 * Java Spring Boot equivalent of the TypeScript SafeTrackClient.
 * This component provides convenient methods to call the backend REST API.
 * It mirrors the TypeScript methods, using RestTemplate for HTTP requests.
 */
@Component
public class SafeTrackClient {

    private final RestTemplate restTemplate;
    private final String baseUrl;
    private String token;

    public SafeTrackClient(@Value("${safetrack.base-url:http://localhost:4000}") String baseUrl) {
        this.baseUrl = baseUrl;
        this.restTemplate = new RestTemplate();
    }

    public void setToken(String token) {
        this.token = token;
    }

    // ---------- Auth ----------
    public AuthResponse register(RegisterRequest body) {
        return postJson("/api/auth/register", body, AuthResponse.class);
    }

    public AuthResponse login(LoginRequest body) {
        return postJson("/api/auth/login", body, AuthResponse.class);
    }

    public HealthResponse health() {
        return getJson("/api/health", HealthResponse.class);
    }

    // ---------- Incidents ----------
    public List<Incident> incidents() {
        return getJsonList("/api/incidents", Incident[].class);
    }

    public Incident createIncident(MultiValueMap<String, Object> form) {
        return postForm("/api/incidents", form, Incident.class);
    }

    public Incident updateIncident(String id, Map<String, Object> updates) {
        return patchJson(String.format("/api/incidents/%s", id), updates, Incident.class);
    }

    public SuccessResponse deleteIncident(String id) {
        return delete(String.format("/api/incidents/%s", id), SuccessResponse.class);
    }

    // ---------- Routes ----------
    public List<SavedRoute> routes() {
        return getJsonList("/api/routes", SavedRoute[].class);
    }

    public SavedRoute saveRoute(Map<String, Object> body) {
        return postJson("/api/routes", body, SavedRoute.class);
    }

    public SavedRoute toggleFavorite(String id) {
        return patch(String.format("/api/routes/%s/favorite", id), null, SavedRoute.class);
    }

    public SuccessResponse deleteRoute(String id) {
        return delete(String.format("/api/routes/%s", id), SuccessResponse.class);
    }

    // ---------- Chat ----------
    public ChatResult chat(String message, List<Map<String, String>> history) {
        var payload = Map.of("message", message, "history", history);
        return postJson("/api/ai/chat", payload, ChatResult.class);
    }

    public List<ChatHistoryItem> chatHistory() {
        return getJsonList("/api/ai/history", ChatHistoryItem[].class);
    }

    public SuccessResponse clearChatHistory() {
        return delete("/api/ai/history", SuccessResponse.class);
    }

    // ---------- Music ----------
    public List<MusicTrack> tracks() {
        return getJsonList("/api/music/tracks", MusicTrack[].class);
    }

    public MusicTrack uploadTrack(MultiValueMap<String, Object> form) {
        return postForm("/api/music/tracks", form, MusicTrack.class);
    }

    public SuccessResponse deleteTrack(String id) {
        return delete(String.format("/api/music/tracks/%s", id), SuccessResponse.class);
    }

    public List<Playlist> playlists() {
        return getJsonList("/api/music/playlists", Playlist[].class);
    }

    public Playlist createPlaylist(String name) {
        return postJson("/api/music/playlists", Map.of("name", name), Playlist.class);
    }

    public SuccessResponse addTrackToPlaylist(String playlistId, String trackId, int position) {
        var payload = Map.of("track_id", trackId, "position", position);
        return postJson(String.format("/api/music/playlists/%s/tracks", playlistId), payload, SuccessResponse.class);
    }

    public List<MusicTrack> playlistTracks(String playlistId) {
        return getJsonList(String.format("/api/music/playlists/%s/tracks", playlistId), MusicTrack[].class);
    }

    // ---------- Ads ----------
    public List<Ad> ads() {
        return getJsonList("/api/ads", Ad[].class);
    }

    public List<Ad> nearbyAds(double lat, double lng) {
        return getJsonList(String.format("/api/ads/nearby?lat=%s&lng=%s", lat, lng), Ad[].class);
    }

    public List<Ad> myAds() {
        return getJsonList("/api/ads/mine", Ad[].class);
    }

    public Ad createAd(Map<String, Object> body) {
        return postJson("/api/ads", body, Ad.class);
    }

    public Record checkoutAd(String id) {
        // The original TS uses a generic Record<string, unknown>
        return postJson(String.format("/api/ads/%s/checkout", id), null, Record.class);
    }

    public Ad activateAd(String id) {
        return post(String.format("/api/ads/%s/activate", id), null, Ad.class);
    }

    public SuccessResponse deleteAd(String id) {
        return delete(String.format("/api/ads/%s", id), SuccessResponse.class);
    }

    // ---------- Helper HTTP methods ----------

    private HttpHeaders authHeaders() {
        HttpHeaders headers = new HttpHeaders();
        if (token != null && !token.isEmpty()) {
            headers.setBearerAuth(token);
        }
        return headers;
    }

    private <T> T getJson(String path, Class<T> responseType) {
        HttpEntity<Void> request = new HttpEntity<>(authHeaders());
        ResponseEntity<T> response = restTemplate.exchange(baseUrl + path, HttpMethod.GET, request, responseType);
        return response.getBody();
    }

    private <T> List<T> getJsonList(String path, Class<T[]> arrayType) {
        T[] array = getJson(path, arrayType);
        return List.of(array);
    }

    private <T> T postJson(String path, Object body, Class<T> responseType) {
        HttpHeaders headers = authHeaders();
        headers.setContentType(MediaType.APPLICATION_JSON);
        HttpEntity<Object> request = new HttpEntity<>(body, headers);
        ResponseEntity<T> response = restTemplate.exchange(baseUrl + path, HttpMethod.POST, request, responseType);
        return response.getBody();
    }

    private <T> T patchJson(String path, Object body, Class<T> responseType) {
        HttpHeaders headers = authHeaders();
        headers.setContentType(MediaType.APPLICATION_JSON);
        HttpEntity<Object> request = new HttpEntity<>(body, headers);
        ResponseEntity<T> response = restTemplate.exchange(baseUrl + path, HttpMethod.PATCH, request, responseType);
        return response.getBody();
    }

    private <T> T post(String path, Object body, Class<T> responseType) {
        HttpHeaders headers = authHeaders();
        if (body != null) {
            // Assume form-data when body is MultiValueMap
            headers.setContentType(MediaType.MULTIPART_FORM_DATA);
        }
        HttpEntity<Object> request = new HttpEntity<>(body, headers);
        ResponseEntity<T> response = restTemplate.exchange(baseUrl + path, HttpMethod.POST, request, responseType);
        return response.getBody();
    }

    private <T> T postForm(String path, MultiValueMap<String, Object> form, Class<T> responseType) {
        HttpHeaders headers = authHeaders();
        headers.setContentType(MediaType.MULTIPART_FORM_DATA);
        HttpEntity<MultiValueMap<String, Object>> request = new HttpEntity<>(form, headers);
        ResponseEntity<T> response = restTemplate.exchange(baseUrl + path, HttpMethod.POST, request, responseType);
        return response.getBody();
    }

    private <T> T delete(String path, Class<T> responseType) {
        HttpEntity<Void> request = new HttpEntity<>(authHeaders());
        ResponseEntity<T> response = restTemplate.exchange(baseUrl + path, HttpMethod.DELETE, request, responseType);
        return response.getBody();
    }

    private <T> T patch(String path, Object body, Class<T> responseType) {
        HttpHeaders headers = authHeaders();
        headers.setContentType(MediaType.APPLICATION_JSON);
        HttpEntity<Object> request = new HttpEntity<>(body, headers);
        ResponseEntity<T> response = restTemplate.exchange(baseUrl + path, HttpMethod.PATCH, request, responseType);
        return response.getBody();
    }

    // ---------- DTO classes ----------
    public static class User {
        public String id;
        public String name;
        public String email;
        public String avatar_url;
    }

    public static class AuthResponse {
        public String token;
        public User user;
    }

    public static class Incident {
        public String id;
        public String user_id;
        public String type;
        public String title;
        public String description;
        public double latitude;
        public double longitude;
        public String severity;
        public String status;
        public String media_url;
        public String reporter_name;
        public String created_at;
        public String updated_at;
    }

    public static class SavedRoute {
        public String id;
        public String user_id;
        public String name;
        public String origin_name;
        public String destination_name;
        public double origin_lat;
        public double origin_lng;
        public double destination_lat;
        public double destination_lng;
        public Object route_data;
        public boolean is_favorite;
        public String created_at;
    }

    public static class ChatResult {
        public String text;
        public Object action;
    }

    public static class ChatHistoryItem {
        public String role;
        public String content;
        public String created_at;
    }

    public static class MusicTrack {
        public String id;
        public String user_id;
        public String title;
        public String artist;
        public String album;
        public Integer duration;
        public String file_url;
        public String cover_url;
        public String created_at;
    }

    public static class Playlist {
        public String id;
        public String user_id;
        public String name;
        public String cover_url;
        public Integer track_count;
        public String created_at;
    }

    public static class Ad {
        public String id;
        public String user_id;
        public String business_name;
        public String description;
        public double latitude;
        public double longitude;
        public double radius_km;
        public String image_url;
        public String website_url;
        public String payment_status;
        public Boolean active;
        public Integer duration_days;
        public String expires_at;
        public String created_at;
        public Double distance_km;
    }

    public static class SuccessResponse {
        public boolean success;
    }

    // Generic record for checkout response (unknown fields)
    public static class Record extends java.util.HashMap<String, Object> {}

    // Request payload classes
    public static class RegisterRequest {
        public String name;
        public String email;
        public String password;
    }

    public static class LoginRequest {
        public String email;
        public String password;
    }

    public static class HealthResponse {
        public String status;
        public String timestamp;
    }
}
